feat: make Request object thenable

pablopalacios opened the pull request:
Now the request object returned by fetchr crud methods are [thenable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#thenables) by default. 

To make it possible, I had to refactor `httpRequest` to transform it in a class to make it easier to expose `then`, `catch` and `abort` to the users. By doing it, it was also possible to simplify the `Request` class and remove some workarounds related to `abort`.

This PR also implement some of the stuff we discussed in https://github.com/yahoo/fetchr/pull/263#issuecomment-881089203 without the need for a major version bump. 

## Before
```js
// superagent style
fetchr
  .read('foo')
  .end()
  .then()

// callback style without callback
fetchr
  .read('foo', null) // passing a second params was necessary to not trigger the superagent API.
  .then() 
```
## After
```js
// superagent or callback
fetchr
  .read('foo')
  .then()
```

This is also quite nice when doing fire and forget calls with async/await:

```js
await fetchr.update('counter');
```

<!-- The following line must be included in your pull request -->
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

redonkulus commented on README.md:
```diff
@@ -106,19 +106,22 @@ Fetcher.registerService(myDataService);
-        .params({id: ###})
-        .end(function (err, data, meta) {
-        // handle err and/or data returned from data fetcher in this callback
+        .params({ id: 42 })
```
This is more of a style comment and less about the code changes, but I've never been a fan of this programming style. All of this code was written before async / await was available. I would prefer if the API were something simpler like:

```js
try { 
  // read(/* service */, /* params */, /* config */);
  const data = await fetcher.read('data_service', { id: 42 });
} catch (e) {
  console.log(e);
}
```

Maybe this is something we change for a 1.x release.

pablopalacios replied:
Yup, I agree. I only fixed the docs because there are people at my company requiring a more up to date docs. I've also revisited what we discussed on #275 and this is what I want to tackle next. I also think it's possible to make it backward compatible so we can adjust everything before going to v1 and remove all the other APIs.

redonkulus replied:
Makes sense to me. BC is ideal but not a blocker if it makes the API better and user friendly.

redonkulus commented on libs/util/httpRequest.js:
```diff
@@ -9,154 +9,161 @@
 
-function io(options, controller) {
+function _fetch() {
+    var self = this;
```
nit: do we still need to use `self`?

pablopalacios replied:
We need it later in the fetch call itself. Most of the time we could use `this` and only self in the fetch promise. But having both would be annoying so I opted to only use one. I was also very tempted to use arrow functions so we wouldn't need this workaround. I was afraid that you would consider it a breaking change, so I kept the es5 style.

redonkulus commented on libs/util/httpRequest.js:
```diff
@@ -9,154 +9,161 @@
 }
 
-function delayPromise(fn, delay) {
+// _retry is the onReject promise callback that we attach to the
```
Any thing to review in these changes or is it just reorganizing the code?

pablopalacios replied:
Not really, just reorganizing.

redonkulus reviewed (commented):
Overall this looks ok. I'm assuming there are not breaking changes here and just organization and code improvements?

pablopalacios commented:
> Overall this looks ok. I'm assuming there are not breaking changes here and just organization and code improvements?

It's mainly refactoring. Calling `.end()` without a callback is still possible but I removed all the tests and added a warning for the users.

pablopalacios commented:
@redonkulus any blocker?

redonkulus commented:
Published https://github.com/yahoo/fetchr/releases/tag/v0.7.11
